### ISSUE TITLE
[feature] CUE support 

### ISSUE THREAD

**Issue by iyefrat** [moonrepo/moon#1599]
**Is your feature request related to a problem? Please describe.**

From https://github.com/moonrepo/moon/pull/1598 and https://github.com/moonrepo/moon/issues/1569#issuecomment-2245765700 I see that you're planning on migrating the config files from yaml to pkl. We've been using [CUE](https://cuelang.org/) as our source-of-truth configuration format in our monorepo, for `moon.yml`, `package.json`, etc. Have you considered CUE before settling on pkl? Is supporting both an option?

**Describe the solution you'd like**

Using CUE directly to configure moon. I think that it has a strong set of core abstractions behind it, in particular https://cuelang.org/docs/concept/the-logic-of-cue/

**Describe alternatives you've considered**

I haven't yet looked into pkl deeply, and currently we have a manual build step, but the fact that we generate the moon files themselves does cause some paper cuts with things like caching and package installation

**Comment by iyefrat** [comment]
See also: Pulumi supports CUE, and essentially anything that compiles to yaml, via a `compile` field that lets you specify how to generate the yaml
https://www.pulumi.com/blog/extending-pulumi-languages-with-yaml-cue-jsonnet-rust/

**Comment by milesj** [comment]
I haven't used cue really but the main reasons I'm looking forward to Pkl is:

- It is owned by a large company. Apple in this case.
- It is backed by a schema (kind of like JSON schema). 
- It is programmable. Each data type has methods to mutate/operate with.
- The base config file isn't something like YAML or JSON, it's Pkl.
- I just like the syntax. It's kind of a mix of TOML, Terraform, and Gradle.

I don't really plan to support more than 1 config format, so if Pkl is a success, YAML will be removed entirely in v2.

**Comment by milesj** [comment]
@iyefrat I have no idea what pulumi is or what it does, but after a brief glance, I don't see how this would work for moon? moon isn't infrastructure, and a lot of pulumi's APIs are infra centric. 

It also seems like a maintenance burden on me (when something breaks) if users are configuring things in whatever format they want. Most support requests are based around config issues.

**Comment by iyefrat** [comment]
Oh sorry i think i was vague with my wording - I meant the approach linked here https://github.com/moonrepo/moon/issues/1599#issuecomment-2266932875, where you support yaml but let users specify how they want to compile to yaml (via pkl for example), and delegate features like inheritance to the compilation step.

on the moon level you only support the yaml, anything inheritance based is out of scope and delegated to pkl/cue/whatever support.

The main reason I'm concerned about this change is that we use cue to in our monorepo to solve the same class of problems you want to solve here: to define variables, allow shared logic across workspace, not have to reinvent inheritance logic, etc. The issue is that if moon moves to only supporting pkl, we're essentially forced to migrate away from moon, or migrate the monorepo from cue to pkl, and we don't want to be forced in to either of those moves for that reason.

**Comment by milesj** [comment]
> where you support yaml but let users specify how they want to compile to yaml (via pkl for example)

Pkl wouldn't be used to compile to YAML, it would just be Pkl itself. I'm looking for a new configuration format that is stand-alone, and doesn't have a compilation step or intermediary format, but has far more functionality than static formats (JSON, YAML). Something like Starlark is _too_ much, so Pkl feels like a good choice at the moment. I'm not 100% sold on it though, it's still experimental.

> The main reason I'm concerned about this change is that we use cue to in our monorepo to solve the same class of problems you want to solve here: to define variables, allow shared logic across workspace, not have to reinvent inheritance logic, etc. 

It looks like CUE doesn't even support Rust, so that means it's immediately off the table.

**Comment by sharpepd** [comment]
@milesj I am trying to find a feature or discussion on the topic of pkl/configuration language, where you can share your thoughts on this topic and get feedback on the experimental pkl feature. I can see a lot of potential in using Pkl, but for our users, it is "yet another language" to learn. The starlark language looks more attractive to use, because of the familiar syntax to Python developers (and we have a number of things already that are python based). So, it would be interesting to have a place where we can see your reasoning behind statements like "starlark is too much".

**Comment by chadrik** [comment]
I had a look at pkl and CUE and read Miles's [blog post](https://moonrepo.dev/blog/moon-v1.29#experimental-support-for-pkl-based-configuration) on the subject, and here's my POV:

pkl is an interesting and powerful tool, but I would hate for it to be the only way to define moon configs.   One of the things I really like about `moon`, and what sets it apart from many of its competitors, is its streamlined and readable config files.  I manage a team that has a monorepo with 40+ developers contributing to it, ranging in skill from junior to senior.  I want any of them to be able to browse the task configuration and get a sense of how it works, and the vast majority of them already understand the relatively simple semantics of YAML. 

Yes, there are cases where a for loop inside my config file would be a powerful way to reduce boilerplate, but that benefit does not outweigh the trade offs of requiring my developers spend hours reading the pkl docs to understand nuances like the difference between a `List` and `Listing` just to make a small tweak to our task system. 

Let's compare a few options, starting with the example from your blog post:

`.moon/tasks.pkl`
```pkl
tasks {
  for (_os in List("linux", "macos", "windows")) {
    ["build-\(_os)"] {
      command = "cargo"
      args = List(
        "--target",
        if (_os == "linux") "x86_64-unknown-linux-gnu"
          else if (_os == "macos") "x86_64-apple-darwin"
          else "i686-pc-windows-msvc",
        "--verbose"
      )
      options {
        os = _os
      }
    }
  }
}
```

Moon already supports templating with tera, so adding support for reading templated config files directly could be a fine compromise:

`.moon/tasks.yml.tera`

```yml
tasks:
  {%- for os in ["linux", "macos", "windows"] %}
  build-{{ os }}:
    command: "cargo"
    args:
      - "--target"
      {%- if os == "linux" %}
      - "x86_64-unknown-linux-gnu"
      {%- elif os == "macos" %}
      - "x86_64-apple-darwin"
      {%- else %}
      - "i686-pc-windows-msvc"
      {%- endif %}
      - "--verbose"
    options:
      os: {{ os }}
  {%- endfor %}
```

But honestly, in many cases I would just go with the explicit approach to keep it simple:

`.moon/tasks.yml`
```yaml
tasks:
  build-linux:
    command: "cargo"
    args: ["--target=x86_64-unknown-linux-gnu", "--verbose"]
    options:
      os: linux

  build-macos:
    command: "cargo"
    args: ["--target=x86_64-apple-darwin", "--verbose"]
    options:
      os: macos

  build-windows:
    command: "cargo"
    args: ["--target=i686-pc-windows-msvc", "--verbose"]
    options:
      os: windows
```

tl;dr  I'm fine with pkl as an officially supported config format, but please don't take away YAML.  I think this would be a deal-breaker for me and my team.   If one of the benefits of pkl is schema-definition and validation, can't you use [pkl libraries to load yaml](https://pkl-lang.org/package-docs/pkl/0.26.1/yaml/Parser.html) and easily support both formats?

**Comment by milesj** [comment]
I've thought about Tera before, but the primary reason I won't support it is that it will break syntax highlighting in editors, and make it nigh impossible for consumers to parse these files manually (I know people are doing this).

**Comment by milesj** [comment]
Added support for JSON/TOML/HCL in the upcoming v2. No plans for CUE since there's no Rust crate available from what I can tell.

- Alpha is available now: https://moonrepo.dev/blog/moon-v2-alpha
- More about v2: https://moonrepo.dev/blog/moon-v2.0
